File: connectors/clients/julia/src/octad.jl
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
#
# VeriSimDB Julia Client — Octad CRUD operations.
#
# This file provides create, read, update, delete, and paginated list
# operations for VeriSimDB octad entities. All functions communicate with
# the VeriSimDB REST API via the Client's HTTP helpers.

"""
    create_octad(client::Client, input::OctadInput) -> Octad

Create a new octad on the VeriSimDB server.

# Arguments
- `client::Client` — The authenticated client.
- `input::OctadInput` — The octad input describing modalities and data.

# Returns
The newly created `Octad` with server-assigned ID and timestamps.

# Throws
`VeriSimError` on HTTP or server failure.
"""
function create_octad(client::Client, input::OctadInput)::Octad
    resp = do_post(client, "/api/v1/octads", input)
    return parse_response(Octad, resp)
end

"""
    get_octad(client::Client, id::String) -> Octad

Retrieve a single octad by its unique identifier.

# Arguments
- `client::Client` — The authenticated client.
- `id::String` — The octad's unique identifier.

# Returns
The requested `Octad`.

# Throws
`NotFoundError` if the octad does not exist.
"""
function get_octad(client::Client, id::String)::Octad
    resp = do_get(client, "/api/v1/octads/$id")
    return parse_response(Octad, resp)
end

"""
    update_octad(client::Client, id::String, input::OctadInput) -> Octad

Update an existing octad with the given input fields.
Only the fields present in the input are modified; others remain unchanged.

# Arguments
- `client::Client` — The authenticated client.
- `id::String` — The octad's unique identifier.
- `input::OctadInput` — The fields to update.

# Returns
The updated `Octad`.

# Throws
`VeriSimError` on failure.
"""
function update_octad(client::Client, id::String, input::OctadInput)::Octad
    resp = do_put(client, "/api/v1/octads/$id", input)
    return parse_response(Octad, resp)
end

"""
    delete_octad(client::Client, id::String) -> Bool

Delete a octad by its unique identifier.

# Arguments
- `client::Client` — The authenticated client.
- `id::String` — The octad's unique identifier.

# Returns
`true` if the octad was successfully deleted.

# Throws
`VeriSimError` on failure.
"""
function delete_octad(client::Client, id::String)::Bool
    resp = do_delete(client, "/api/v1/octads/$id")
    status = resp.status
    if status == 204 || status == 200
        return true
    end
    throw(error_from_status(status, String(resp.body)))
end

"""
    list_octads(client::Client; page::Int=1, per_page::Int=20) -> PaginatedResponse

Retrieve a paginated list of octads.

# Keyword Arguments
- `page::Int` — Page number (1-indexed). Defaults to 1.
- `per_page::Int` — Number of octads per page. Defaults to 20.

# Returns
A `PaginatedResponse` containing octads and pagination metadata.

# Throws
`VeriSimError` on failure.
"""
function list_octads(client::Client; page::Int=1, per_page::Int=20)::PaginatedResponse
    resp = do_get(client, "/api/v1/octads?page=$page&per_page=$per_page")
    return parse_response(PaginatedResponse, resp)
end

File: connectors/clients/rust/src/octad.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>

//! Octad CRUD operations.
//!
//! Octads are the fundamental multi-modal entities in VeriSimDB. Each octad can
//! carry data across all eight modalities (graph, vector, tensor, semantic,
//! document, temporal, provenance, spatial). This module provides create, read,
//! update, delete, and list operations as methods on [`VeriSimClient`].

use crate::client::VeriSimClient;
use crate::error::Result;
use crate::types::{Octad, OctadInput, PaginatedResponse};

impl VeriSimClient {
    /// Create a new octad entity.
    ///
    /// The server assigns a UUID and timestamps; the returned [`Octad`] contains
    /// the fully-populated record.
    ///
    /// # Arguments
    ///
    /// * `input` — The octad payload. At minimum, `name` should be set.
    ///
    /// # Errors
    ///
    /// Returns [`VeriSimError::Validation`] if required fields are missing, or a
    /// network / server error on transport failure.
    pub async fn create_octad(&self, input: &OctadInput) -> Result<Octad> {
        self.post("/api/v1/octads", input).await
    }

    /// Retrieve a single octad by its unique identifier.
    ///
    /// # Errors
    ///
    /// Returns [`VeriSimError::NotFound`] if no octad exists with the given `id`.
    pub async fn get_octad(&self, id: &str) -> Result<Octad> {
        let path = format!("/api/v1/octads/{id}");
        self.get(&path).await
    }

    /// Update an existing octad entity.
    ///
    /// Only the fields present in `input` are modified; omitted fields retain
    /// their current values (partial update / merge semantics).
    ///
    /// # Errors
    ///
    /// Returns [`VeriSimError::NotFound`] if the octad does not exist.
    pub async fn update_octad(&self, id: &str, input: &OctadInput) -> Result<Octad> {
        let path = format!("/api/v1/octads/{id}");
        self.put(&path, input).await
    }

    /// Delete a octad entity by its unique identifier.
    ///
    /// This is a hard delete — the entity and all associated modality data are
    /// removed. Provenance records are retained for auditability.
    ///
    /// # Errors
    ///
    /// Returns [`VeriSimError::NotFound`] if the octad does not exist.
    pub async fn delete_octad(&self, id: &str) -> Result<()> {
        let path = format!("/api/v1/octads/{id}");
        self.delete(&path).await
    }

    /// List octad entities with pagination.
    ///
    /// # Arguments
    ///
    /// * `limit`  — Maximum number of results to return (server may cap this).
    /// * `offset` — Zero-based offset for pagination.
    ///
    /// # Returns
    ///
    /// A [`PaginatedResponse`] containing the requested page of octads.
    pub async fn list_octads(
        &self,
        limit: usize,
        offset: usize,
    ) -> Result<PaginatedResponse<Octad>> {
        let path = format!("/api/v1/octads?limit={limit}&offset={offset}");
        self.get(&path).await
    }
}
